src/sema/unused_variable.rs

```diff
@@ -228,7 +229,9 @@ pub fn check_var_usage_expression(
 pub fn emit_warning_local_variable(variable: &symtable::Variable) -> Option<Diagnostic> {
     match &variable.usage_type {
         VariableUsage::Parameter => {
-            if !variable.read {
+            if !variable.read
+                && !matches!(variable.storage_location, Some(StorageLocation::Storage(_)))
+            {
```

Comment:
This contact has a function like:
```
            function bar(int[] storage foo) internal {
                foo[0] = 102;
            }
```
and solang says `foo` is unread, which is wrong of course.

However your PR suppresses the warning for `storage` parameters. That's not right, what if we have a function like:
```
            function bar(int[] storage foo) internal {
            }
```
We still want a warning then, right?

Reply:
Yes. I've fixed this.

Reply:
I'm being pedantic now, but this is hard to read. How about:
```
if !variable.read {
    if !variable.assigned {
       // warn never read/assigned
    } else if !storage {
       // warn never been read
    }
}
```
I am not sure why there is special case for storage variables though.  This needs a comment

Reply:
> I am not sure why there is special case for storage variables though. 

A function like this:
```solidity
function foo(int[] storage bar) {
int[2] = 5;
}
```
should not have a warning about `bar` being unread, because it is a reference to the storage.

Reply:
why not?

Reply:
```solidity

contract test {
int[] vec;
function foo(int[] storage bar) {
    bar[2] = 5;
}

function foo2() returns (int) {
   vec.push(1);
   vec.push(2);
   vec.push(3);
   foo(vec);
   return vec[2];
}

}
```

We would raise a warning about an unread variable in this case, but it is not actually unread, just because the argument is a reference to the storage and not the value itself.

Reply:
Sorry, yes you are right. However, what about the non-storage version of that:

```solidity
contract test {
function foo(int[] bar) pure {
    bar[2] = 5;
}

function foo2() public returns (int) {
int[] vec = new int[](0);
   vec.push(1);
   vec.push(2);
   vec.push(3);
   foo(vec);
   return vec[2];
}

}
```

Why are we generating a warning for that?

Also in this following case the arguments are never used.

```solidity
contract c {
	int[] var;

	function g(int[] storage x) internal {
		x = var;
	}
	function gg(int[] x) internal {
		x = new int[](2);
	}

}
```

Reply:
> Why are we generating a warning for that?

The reaching definitions implementation does not work across function calls. This is a limitation of the way we detect unused values.

> Also in this following case the arguments are never used.

There is a warning for `gg`, because `x` is never read. For `g` there is no warning, unfortunately.